profiling/plugins/recorder_plugin/recorder_plugin.c

```diff
@@ -0,0 +1,741 @@
+  }
+  return dest;
+}
+
+/* thread_id will point to thread_id_v if the thread is created successfully;
+ * null otherwise.
+ */
+static uv_thread_t thread_id_v, *thread_id = NULL;
+static datadog_php_channel channel;
+static ddprof_ffi_ProfileExporterV3 *exporter = NULL;
+
+typedef struct record_msg_s record_msg;
+struct record_msg_s {
+  datadog_php_record_values record_values;
+  int64_t thread_id;
+  datadog_php_stack_sample sample;
+};
+
+DDTRACE_HOT __attribute__((nonnull)) bool
+datadog_php_recorder_plugin_record(datadog_php_record_values record_values,
+                                   int64_t tid,
+                                   const datadog_php_stack_sample *sample) {
+  if (!enabled) {
+    const char *str =
+        "[Datadog Profiling] Sample dropped because profiling has been disabled.";
+    datadog_php_string_view msg = {strlen(str), str};
+    prof_logger.log(DATADOG_PHP_LOG_WARN, msg);
+    return false;
+  }
+
+  record_msg *message = malloc(sizeof(record_msg));
+  if (message) {
+    message->record_values = record_values;
+    message->sample = *sample;
+    message->thread_id = tid;
+
+    bool success = channel.sender.send(&channel.sender, message);
+    if (!success) {
+      // todo: is this too noisy even for debug?
+      const char *str =
+          "[Datadog Profiling] Failed to store sample for aggregation; queue is likely full or closed.\n";
+      datadog_php_string_view msg = {strlen(str), str};
+      prof_logger.log(DATADOG_PHP_LOG_DEBUG, msg);
+      free(message);
+    }
+    return success;
+  }
+  return false;
+}
+
+typedef struct instant_s instant;
+struct instant_s {
+  // private:
+  uint64_t started_at_nanos;
+};
+
+static uint64_t instant_elapsed(instant self) {
+  return (uv_hrtime() - self.started_at_nanos);
+}
+
+static instant instant_now(void) {
+  instant now = {.started_at_nanos = uv_hrtime()};
+  return now;
+}
+
+static bool ddprof_ffi_export(datadog_php_static_logger *logger,
+                              const struct ddprof_ffi_Profile *profile) {
+  struct ddprof_ffi_EncodedProfile *encoded_profile =
+      ddprof_ffi_Profile_serialize(profile);
+  if (!encoded_profile) {
+    logger->log_cstr(DATADOG_PHP_LOG_WARN,
+                     "[Datadog Profiling] Failed to serialize profile.");
+    return false;
+  }
+
+  ddprof_ffi_Timespec start = encoded_profile->start;
+  ddprof_ffi_Timespec end = encoded_profile->end;
+
+  ddprof_ffi_Buffer profile_buffer = {
+      .ptr = encoded_profile->buffer.ptr,
+      .len = encoded_profile->buffer.len,
+      .capacity = encoded_profile->buffer.capacity,
+  };
+
+  ddprof_ffi_File files_[] = {{
+      .name = to_byteslice("profile.pprof"),
+      .file = &profile_buffer,
+  }};
+
+  struct ddprof_ffi_Slice_file files = {.ptr = files_,
+                                        .len = sizeof files_ / sizeof *files_};
+  ddprof_ffi_Request *request =
+      ddprof_ffi_ProfileExporterV3_build(exporter, start, end, files, 10000);
```

Comment:
Nit: When the API exists in a separate repo, it may be non-trivial to look up what this magic number is (`timeout_ms: u64`). Would you be open to switching out the literal with a const or macro? And perhaps add some context as to why `10000` was chosen as the timeout?

Reply:
Great points. I'll extract it as a parameter to `ddprof_ffi_export` (which contains the quoted snippet) so the caller can choose the timeout, and document the choice of timeout there.